### ISSUE TITLE
Per-model versioning without polymorphic associations

### ISSUE THREAD

**Issue by bdmac** [paper-trail-gem/paper_trail#1104]
**Is your feature suggestion related to a problem? Please describe.**

Your docs suggest that in some cases it is preferable to store the versions in their own tables for performance reasons. We also want to do this for referential integrity reasons since polymorphic tables make this difficult. I can see how to do this but my problem is the `PaperTrail::VersionConcern` that needs to be included makes a ton of assumptions about the structure of the table essentially being polymorphic since it automatically adds `belongs_to :item, polymorphic: true` which then assumes `item_id` and `item_type` fields exist.

If you are using one table per versioned model (e.g. `ContactVersions`) there is no reason to have `item_id` and `item_type` there instead of simply `contact_id`. I can get around most of the item requirements (I hope) in my base class by overriding the various `item` methods and using `ActiveRecord::Reflections` and the `BelongsToAssociation` class to get at the (in this case) `belongs_to :contact` association.

The `PaperTrail::VersionConcern` though automatically adds that `item` relationship which is completely unnecessary in this case. I cannot see a way to have it _not_ add that association nor can I find a way to basically _remove_ that association after including the concern.

**Describe the solution you'd like to build**

An option that would allow per-model versioning tables to not use polymorphic relationships unnecessarily.

**Describe alternatives you've considered**

I'm deep in the middle of trying to hack around the automatically added `item` relationship.

**Comment by jaredbeck** [comment]
> .. to store the versions in their own tables .. We also want to do this for referential integrity ..

Nice, I'm a big fan of foreign key constraints.

> .. my problem is the PaperTrail::VersionConcern that needs to be included makes a ton of assumptions about the structure of the table essentially being polymorphic since it automatically adds belongs_to :item, polymorphic: true which then assumes item_id and item_type fields exist.

We could have two modules, `VersionConcernPolymorphic` and `VersionConcernMonomorphic`. They could both inherit from `VersionConcern`.

**Comment by jaredbeck** [comment]
> The has_paper_trail macro would need to accept an option to disable polymorphism there as well to prevent it from setting up has_many :versions, as: :item (dropping the as: :item portion).

Should that option simply cause `as: :item` to be omitted, or should it cause the `has_many` call to be omitted entirely? The latter could support future use cases.

**Comment by bdmac** [comment]
That's what I meant by dropping the `as: :item`. Yah it should just be omitted leaving the `has_many :items` I think.

**Comment by jaredbeck** [comment]
> .. should it cause the has_many call to be omitted entirely? The latter could support future use cases.

What is your analysis of this suggestion?

**Comment by jaredbeck** [comment]
>>> The has_paper_trail macro would need to accept an option to disable polymorphism there as well to prevent it from setting up has_many :versions, as: :item (dropping the as: :item portion).

>> Should that option simply cause as: :item to be omitted, or should it cause the has_many call to be omitted entirely? The latter could support future use cases.

> I can't follow when you would not want the has_many association to be there at all. Then you would not be able to do model.versions or use most of the functionality provided by this gem right?

You have a class like `ContactVersion`, right? You would be responsible for calling `has_many` there. That way, you have ultimate flexibility. You can omit `as: :item` if you want, or you can do anything else your heart desires :grin:

**Comment by bdmac** [comment]
Well that's backwards though. `ContactVersion belongs_to :contact`. I'm talking about the `has_paper_trail` call that causes those versions to be created which is added to the `Contact` model directly. This macro would need to not assume polymorphism by accepting an option to skip the `as: :item` part of the relationship.

**Comment by jaredbeck** [comment]
> Well that's backwards though.

Oops, yeah. So you would be responsible for calling has_many in your `Contact` class.

> That way, you have ultimate flexibility. You can omit as: :item if you want, or you can do anything else your heart desires 😁

What do you think?

**Comment by bdmac** [comment]
It feels brittle. You still have to call `has_paper_trail` to set up all the various model callbacks and if you _forget_ to add the `has_many :versions` or call it something PaperTrail doesn't expect then I don't think anything would work would it?

**Comment by jaredbeck** [comment]
>> So you would be responsible for calling has_many in your Contact class.

> It feels brittle. You still have to call has_paper_trail to set up all the various model callbacks and if you forget to add the has_many :versions or call it something PaperTrail doesn't expect then I don't think anything would work would it?

Correct. I guess my concern is that there are a lot of parts of `has_many` that people might want to customize in the future, and I don't want to have a different config option in PT for each of them.

But, how many config options could we end reasonably end up with? We already have two: versions_association_name and version_class_name. I can see people wanting to customize or omit the `order`. That'd be three options. Omitting `as: :item` makes four. Someone might want `counter_cache`, `dependent`, maybe `validate`, maaaybe `extend`. That'd be eight options max. 

Not only would eight options be a lot of code to add to PT, it's also this whole layer that we're making people learn, when they already know how to use `has_many`. Is the extra layer helping?

**Comment by bdmac** [comment]
I mean couldn't the macro method just pass everything through to the `has_many` with defaults set appropriately? Something like:

```
def has_paper_trail(**args)
  # Lots of other stuff
  # Set lots of defaults for args like as: :item
  has_many **args
end
```

**Comment by jaredbeck** [comment]
> Something like:
> 
> ```
> def has_paper_trail(**args)
>   has_many **args
> end
> ```

I like it. What would you call the `has_paper_trail` option? `versions_association_options`? I think we'll want to deprecate `versions_association_name`. What should we do with `version_class_name`?

**Comment by jaredbeck** [comment]
Brian, are you still working on this? Want me to keep it open?
